File: python/pyES/Synapps.py

import Common

class Region( object ) :

    @classmethod
    def create( c, params = None ) :
        if params :
            return Region( **params )
        else :
            return Region( None, None, None, None )

    def __init__( self, apply, weight, lower, upper ) :
        self.apply  = apply
        self.weight = weight
        self.lower  = lower
        self.upper  = upper

class Evaluator( object ) :

    @classmethod
    def create( c, params = None ) :
        if not params :
            return Evaluator( None, 2, [] )
        regions = []
        for apply, weight, lower, upper in zip( *[ params[ "regions" ][ attr ] for attr in "apply weight lower upper".split() ] ) :
            regions.append( Region( apply, weight, lower, upper ) )
        params[ "regions" ] = regions
        return Evaluator( **params )

    def __init__( self, target_file, vector_norm, regions ) :
        self.target_file = target_file
        self.vector_norm = vector_norm
        self.regions     = regions

    def __repr__( self ) :
        output =  "evaluator :\n"
        output += "    %-12s : %s\n" % ( "target_file", self.target_file )
        output += "    %-12s : %s\n" % ( "vector_norm", self.vector_norm )
        output += "    %-12s :\n" % "regions"
        for attr in "apply weight lower upper".split() :
            output += "        %-8s : [ " % attr
            for region in self.regions :
                value = getattr( region, attr )
                if type( value ) is bool :
                    output += "%8s," % ( "Yes" if value else "No" )
                else :
                    output += "%8.2f," % value
            output = output.rstrip( "," )                
            output += " ]\n"
        return output.rstrip()

class Variable( object ) :

    @classmethod
    def create( c, params = None ) :
        if params :
            return Variable( **params )
        else :
            return Variable( False, 1.0, 0.0, 10.0, 1.0 )

    def __init__( self, fixed, start, lower, upper, scale ) :
        self.fixed = fixed
        self.start = start
        self.lower = lower
        self.upper = upper
        self.scale = scale

class Ion( object ) :

    @classmethod
    def create( c, params = None ) :
        if params :
            return Ion( **params )
        else :
            return Ion( None, None, None, Variable.create(), Variable.create(), Variable.create(), Variable.create(), Variable.create() )

    def __init__( self, ion, active, detach, log_tau, v_min, v_max, aux, temp ) :
        self.ion     = ion
        self.active  = active
        self.detach  = detach
        self.log_tau = log_tau
        self.v_min   = v_min
        self.v_max   = v_max
        self.aux     = aux
        self.temp    = temp

class Config( object ) :

    @classmethod
    def create( c, params = None ) :
        if not params :
            return Config( None, None, 
                    Variable.create(), Variable.create(), Variable.create(), Variable.create(), Variable.create(), Variable.create(), [] )
        for var_name in "a0 a1 a2 v_phot v_outer t_phot".split() :
            params[ var_name ] = Variable.create( params[ var_name ] )
        ions = []
        for i, ion in enumerate( params[ "ions" ] ) :
            ion_params = { "ion" : ion, "active" : params[ "active" ][ i ], "detach" : params[ "detach" ][ i ] }
            for var_name in "log_tau v_min v_max aux temp".split() :
                ion_params[ var_name ] = Variable( *[ params[ var_name ][ attr ][ i ] for attr in "fixed start lower upper scale".split() ] )
            ions.append( Ion.create( ion_params ) )
        for var_name in "ions active detach log_tau v_min v_max aux temp".split() :
            del params[ var_name ]
        params[ "ions" ] = ions
        return Config( **params )
    
    def __init__( self, fit_file, cache_file, a0, a1, a2, v_phot, v_outer, t_phot, ions ) :
        self.fit_file   = fit_file
        self.cache_file = cache_file
        self.a0         = a0
        self.a1         = a1
        self.a2         = a2
        self.v_phot     = v_phot
        self.v_outer    = v_outer
        self.t_phot     = t_phot
        self.ions       = ions

    def __repr__( self ) :
        output =  "config :\n"
        output += "    %-12s : %s\n" % ( "fit_file"  , self.fit_file   )
        output += "    %-12s : %s\n" % ( "cache_file", self.cache_file )
        output += "\n"
        for var_name in "a0 a1 a2 v_phot v_outer t_phot".split() :
            output += "    %-12s : {" % var_name
            for attr in "fixed start lower upper scale".split() :
                value = getattr( getattr( self, var_name ), attr )
                if type( value ) is bool :
                    output += " %5s: %3s," % ( attr, ( "Yes" if value else "No" ) )
                else :
                    output += " %5s: %6.2f," % ( attr, value )
            output = output.rstrip( "," )
            output += " }\n"
        output += "\n"
        for var_name in "ion active detach".split() :
            output += "    %-12s : [ " % ( "ions" if var_name == "ion" else var_name )
            for value in [ getattr( ion, var_name ) for ion in self.ions ] :
                if type( value ) is bool :
                    output += " %6s," % ( "Yes" if value else "No" )
                else :
                    output += " %6s," % value
            output = output.rstrip( "," )
            output += " ]\n"
        output += "\n"
        for var_name in "log_tau v_min v_max aux temp".split() :
            output += "    %-12s :\n" % var_name
            for attr in "fixed start lower upper scale".split() :
                output += "        %-8s : [ " % attr
                for value in [ getattr( getattr( ion, var_name ), attr ) for ion in self.ions ] :
                    if type( value ) is bool :
                        output += " %6s," % ( "Yes" if value else "No" )
                    else:
                        output += " %6.2f," % value
                output = output.rstrip( "," )
                output += " ]\n"
        return output.rstrip()

class Synapps( object ) :

    @classmethod
    def create( c, params = None ) :
        if params :
            grid      = Common.Grid.create    ( params[ "grid"      ] )
            opacity   = Common.Opacity.create ( params[ "opacity"   ] )
            source    = Common.Source.create  ( params[ "source"    ] )
            spectrum  = Common.Spectrum.create( params[ "spectrum"  ] )
            evaluator = Evaluator.create      ( params[ "evaluator" ] )
            config    = Config.create         ( params[ "config"    ] )
            return Synapps( grid, opacity, source, spectrum, evaluator, config )
        else :
            grid      = Common.Grid.create    ()
            opacity   = Common.Opacity.create ()
            source    = Common.Source.create  ()
            spectrum  = Common.Spectrum.create()
            evaluator = Evaluator.create      ()
            config    = Config.create         ()
            return Synapps( grid, opacity, source, spectrum, evaluator, config )

    def __init__( self, grid, opacity, source, spectrum, evaluator, config ) :
        self.grid      = grid
        self.opacity   = opacity
        self.source    = source
        self.spectrum  = spectrum
        self.evaluator = evaluator
        self.config    = config

    def __repr__( self ) :
        output = ""
        for attr in "grid opacity source spectrum evaluator config".split() :
            output += "%s\n" % getattr( self, attr )
        return output.rstrip()

if __name__ == "__main__" :

    import sys
    import yaml

    if sys.argv[ 1 : ] :
        print Synapps.create( yaml.load( open( sys.argv[ 1 ], "r" ) ) )
    else :
        print Synapps.create()

File: src/synapps/synapps.cc
//
// File    : synapps.cc
// --------------------
//
// ES: Elementary Supernova Spectrum Synthesis, Copyright (c) 2010, The
// Regents of the University of California, through Lawrence Berkeley
// National Laboratory (subject to receipt of any required approvals
// from the U.S. Dept. of Energy). All rights reserved.
//
// If you have questions about your rights to use or distribute this
// software, please contact Berkeley Lab's Technology Transfer
// Department at TTD@lbl.gov.
//
// NOTICE. This software was developed under partial funding from the
// U.S. Department of Energy. As such, the U.S. Government has been
// granted for itself and others acting on its behalf a paid-up,
// nonexclusive, irrevocable, worldwide license in the Software to
// reproduce, prepare derivative works, and perform publicly and display
// publicly. Beginning five (5) years after the date permission to
// assert copyright is obtained from the U.S. Department of Energy, and
// subject to any subsequent five (5) year renewals, the U.S. Government
// is granted for itself and others acting on its behalf a paid-up,
// nonexclusive, irrevocable, worldwide license in the Software to
// reproduce, prepare derivative works, distribute copies to the public,
// perform publicly and display publicly, and to permit others to do so.
//

#include "ES_Synapps.hh"
#include "ES_Synow.hh"

#include <appspack/APPSPACK_GCI.hpp>                  // APPSPACK's interface to MPI.
#include <appspack/APPSPACK_Executor_MPI.hpp>         // MPI executor.
#include <appspack/APPSPACK_Solver.hpp>               // Solver.

#include <yaml-cpp/yaml.h>

#include <fstream>
#include <csignal>
#include <csetjmp>

jmp_buf env;

void signal_handler( int param )
{
    longjmp( env, 1 );
}

int main( int argc, char* argv[] )
{

    // Initialize MPI, 2 or more processors.

    int rank = APPSPACK::GCI::init( argc, argv );
    int workers = APPSPACK::GCI::getNumProcs() - 1;
    if( workers < 1 )
    {
        std::cerr << "ERROR: Program requires at least 2 processors." << std::endl;
        APPSPACK::GCI::exit();
        return 137;
    }

    // Configuration in this application comes from a YAML file.

    YAML::Node yaml = YAML::LoadFile( argv [ 1 ] );

    // Target and output spectra.  The output spectrum is sampled
    // at the same wavelengths as the target spectrum.

    std::string target_file = yaml[ "evaluator" ][ "target_file" ].as<std::string>();
    ES::Spectrum target    = ES::Spectrum::create_from_ascii_file( target_file.c_str() );
    ES::Spectrum output    = ES::Spectrum::create_from_spectrum( target );
    ES::Spectrum reference = ES::Spectrum::create_from_spectrum( target );

//  target.rescale_median_flux();

    // Grid object.

    ES::Synow::Grid grid = ES::Synow::Grid::create(
            target.min_wl(),
            target.max_wl(),
            yaml[ "grid"   ][ "bin_width"   ].as<double>(),
            yaml[ "grid"   ][ "v_size"      ].as<int>(),
            yaml[ "grid"   ][ "v_outer_max" ].as<double>() );

    // Opacity operator.

    ES::Synow::Opacity opacity( grid,
            yaml[ "opacity" ][ "line_dir"    ].as<std::string>(),
            yaml[ "opacity" ][ "ref_file"    ].as<std::string>(),
            yaml[ "opacity" ][ "form"        ].as<std::string>(),
            yaml[ "opacity" ][ "v_ref"       ].as<double>(),
            yaml[ "opacity" ][ "log_tau_min" ].as<double>() );

    // Source operator.

    ES::Synow::Source source( grid,
            yaml[ "source" ][ "mu_size" ].as<int>() );

    // Spectrum operator.

    ES::Synow::Spectrum spectrum( grid, output, reference,
            yaml[ "spectrum" ][ "p_size"  ].as<int>(),
            yaml[ "spectrum" ][ "flatten" ].as<bool>() );

    // Evaluator.

    std::vector< int > ions;
    for( size_t i = 0; i < yaml[ "config" ][ "active" ].size(); ++ i )
    {
        if( ! yaml[ "config" ][ "active" ][ i ] ) continue;
        ions.push_back( yaml[ "config" ][ "ions" ][ i ].as<int>() );
    }

    std::vector< double > region_weight;
    std::vector< double > region_lower;
    std::vector< double > region_upper;
    for( size_t i = 0; i < yaml[ "evaluator" ][ "regions" ][ "apply" ].size(); ++ i )
    {
        if( ! yaml[ "evaluator" ][ "regions" ][ "apply" ][ i ].as<bool>() ) continue;
        region_weight.push_back( yaml[ "evaluator" ][ "regions" ][ "weight" ][ i ].as<double>() );
        region_lower.push_back ( yaml[ "evaluator" ][ "regions" ][ "lower"  ][ i ].as<double>() );
        region_upper.push_back ( yaml[ "evaluator" ][ "regions" ][ "upper"  ][ i ].as<double>() );
    }

    ES::Synapps::Evaluator evaluator( grid, target, output, ions, region_weight, region_lower, region_upper,
            yaml[ "evaluator" ][ "vector_norm" ].as<double>() );

    // Master section.

    if( rank == 0 )
    {

        ES::Synapps::Config config( yaml[ "config" ] );

        // Executor, constraints, solver.

        APPSPACK::Executor::MPI       executor;
        APPSPACK::Constraints::Linear linear( config.params.sublist( "Linear" ) );
        APPSPACK::Solver              solver( config.params.sublist( "Solver" ), executor, linear );
//      APPSPACK::Solver::State       state = solver.solve();
        APPSPACK::Solver::State       state;

        // Signal handler, wrapping solver.

        signal( SIGTERM, signal_handler );

        int terminated = setjmp( env );
        if( ! terminated )
        {
            state = solver.solve();
        }
        else
        {
            std::cerr << std::endl;
            std::cerr << "WARNING!" << std::endl;
            std::cerr << "An abnormal exit condition has been reached, probably due to"  << std::endl;
            std::cerr << "a TERM signal.  Attempting to write out the best fit found so" << std::endl;
            std::cerr << "far, but be forewarned that we probably did not converge."     << std::endl;
            std::cerr << std::endl;
        }

        // Problem solved, terminate workers.

        APPSPACK::GCI::initSend();
        APPSPACK::GCI::pack( 1 );
        for( int i = 0; i < workers; ++ i ) APPSPACK::GCI::send( APPSPACK::Executor::MPI::Terminate, i + 1 );

        // Best spectrum.

        int tag;
        APPSPACK::Vector x;
        APPSPACK::Vector f;
        std::string msg;

        x = solver.getBestX();
        f = solver.getBestF();

        evaluator( tag, x, f, msg );

        std::ofstream stream;
        stream.open( config.fit_file.c_str() );
        stream << std::setprecision( 6 ) << output;
        stream.close();

    }

    // Worker section.

    if( rank != 0 )
    {

        while( true )
        {

            // Blocking receive.

            int msg_tag, junk;
            APPSPACK::GCI::recv();
            APPSPACK::GCI::bufinfo( msg_tag, junk );

            // Check for termination.

            if( msg_tag == APPSPACK::Executor::MPI::Terminate ) break;

            // Local workspace.

            int tag;
            APPSPACK::Vector x;
            APPSPACK::Vector f;
            std::string msg;

            // Unpack latest message -- must be done just like this.

            APPSPACK::GCI::unpack( tag );
            APPSPACK::GCI::unpack( x   );

            // Evaluate the function.

            evaluator( tag, x, f, msg );

            // Send reply -- must be done just like this.

            APPSPACK::GCI::initSend();
            APPSPACK::GCI::pack( tag );
            APPSPACK::GCI::pack( f   );
            APPSPACK::GCI::pack( msg );
            APPSPACK::GCI::send( APPSPACK::Executor::MPI::Feval, 0 );

        }

    }

    // Done.

    APPSPACK::GCI::exit();

    return 0;
}
